requestheaders: new parameter inside debug.httpcalls.<BIDDER> to log request header details

ShriprasadM opened the pull request:
httpcalls.\<BIDDER>.**_requestheaders_** - New Debug parameter

@SyntaxNode, Please review and approve this PR.

It will be useful to log a set of request headers sent to the actual Bidder. This will help in verifying if required request headers are correctly getting sent to the actual Bidder.

This parameter will be visible only when query parameter _debug=1_ is set. For example,
> http://localhost:8000/openrtb2/video?debug=1
> <img width="1355" alt="image" src="https://user-images.githubusercontent.com/9314645/105181655-16a09280-5b52-11eb-9a61-1bb377b7d535.png">


In above example **requestheaders** indicating 2 headers are sent to appnexus bidder.

1.  "Accept": ["application/json"],
2. "Content-Type": ["application/json;charset=utf-8"]

****

ShriprasadM commented on exchange/bidder.go:
```diff
@@ -308,17 +308,19 @@ func getAssetByID(id int64, assets []nativeRequests.Asset) (nativeRequests.Asset
+			RequestBody:    string(httpInfo.request.Body),
+			ResponseBody:   string(httpInfo.response.Body),
+			Status:         httpInfo.response.StatusCode,
+			RequestHeaders: httpInfo.request.Headers,
```
Log Request Headers when this is no error

SyntaxNode replied:
Nitpick: Would you be able to move the `RequestHeaders` definition directly below the `RequestBody`?

ShriprasadM replied:
Moved `RequestHeaders ` below `RequestBody`

ShriprasadM replied:
Moved `RequestHeaders ` below `RequestBody`

ShriprasadM commented on exchange/bidder.go:
```diff
@@ -308,17 +308,19 @@ func getAssetByID(id int64, assets []nativeRequests.Asset) (nativeRequests.Asset
-			RequestBody: string(httpInfo.request.Body),
+			Uri:            httpInfo.request.Uri,
+			RequestBody:    string(httpInfo.request.Body),
+			RequestHeaders: httpInfo.request.Headers,
```
Log Request Headers when there is error

ShriprasadM commented on exchange/bidder_test.go:
```diff
@@ -910,6 +916,7 @@ func TestBadResponseLogging(t *testing.T) {
 	if ext.Status != 0 {
 		t.Errorf("The Status code should be 0. Got %d", ext.Status)
 	}
+	assert.Equal(t, info.request.Headers, http.Header(ext.RequestHeaders), "The request headers should be \"header-1:value-1\"")
```
Verify Request headers are logged when there is a bad response

ShriprasadM commented on exchange/bidder_test.go:
```diff
@@ -894,6 +897,9 @@ func TestBadResponseLogging(t *testing.T) {
 		request: &adapters.RequestData{
 			Uri:  "test.com",
 			Body: []byte("request body"),
+			Headers: http.Header{
+				"header-1": []string{"value-1"},
+			},
```
Simulation of request headers set

ShriprasadM commented on exchange/bidder_test.go:
```diff
@@ -918,6 +925,9 @@ func TestSuccessfulResponseLogging(t *testing.T) {
 		request: &adapters.RequestData{
 			Uri:  "test.com",
 			Body: []byte("request body"),
+			Headers: http.Header{
+				"header-1": []string{"value-1", "value-2"},
+			},
```
Simulation of request headers set

ShriprasadM commented on exchange/bidder_test.go:
```diff
@@ -937,6 +947,7 @@ func TestSuccessfulResponseLogging(t *testing.T) {
 	if ext.Status != info.response.StatusCode {
 		t.Errorf("The Status code should be 0. Got %d", ext.Status)
 	}
+	assert.Equal(t, info.request.Headers, http.Header(ext.RequestHeaders), "The request headers should be \"%s\". Got %s", info.request.Headers, ext.RequestHeaders)
```
Verify Request headers are logged when there is a bad response

ShriprasadM commented on exchange/exchangetest/request-multi-bidders-debug-info.json:
```diff
@@ -65,6 +65,7 @@
           {
             "uri": "appnexusTest.com",
             "requestbody": "appnexusTestRequestBody",
+            "requestheaders": null,
```
Added new parameter to satisfy the unit test

ShriprasadM commented on exchange/exchangetest/request-multi-bidders-debug-info.json:
```diff
@@ -95,7 +96,8 @@
             "responsebody": "audienceNetworkTestResponseBody",
-            "status": 200
+            "status": 200,
+            "requestheaders": null
```
Added new parameter to satisfy the unit test

SyntaxNode replied:
Could you please group `requestheaders` closer to `requestbody` in the json files? I think it's easier to read. No need to add it to the end just because it's new.

ShriprasadM replied:
grouped `requestheaders` closer to `requestbody`

ShriprasadM commented on exchange/exchangetest/request-multi-bidders-debug-info.json:
```diff
@@ -150,15 +152,17 @@
               "responsebody": "appnexusTestResponseBody",
-              "status": 200
+              "status": 200,
+              "requestheaders": null
```
Added new parameter to satisfy the unit test

ShriprasadM commented on exchange/exchangetest/request-multi-bidders-debug-info.json:
```diff
@@ -150,15 +152,17 @@
               "responsebody": "audienceNetworkTestResponseBody",
-              "status": 200
+              "status": 200,
+              "requestheaders": null
```
Added new parameter to satisfy the unit test

ShriprasadM commented on openrtb_ext/response.go:
```diff
@@ -55,10 +55,11 @@ type ExtBidderError struct {
+	RequestBody    string              `json:"requestbody"`
+	ResponseBody   string              `json:"responsebody"`
+	Status         int                 `json:"status"`
+	RequestHeaders map[string][]string `json:"requestheaders"`
```
The new parameter **RequestHeaders** will contain the header key and multiple values associated with it.
We have kept value as an array of strings because one header can have multiple values.
Also, this data structure is inline with [http/Header](https://golang.org/pkg/net/http/#Header)

ShriprasadM commented:
@SyntaxNode : Please review and merge

SyntaxNode commented:
> @SyntaxNode : Please review and merge

It's in our queue to review early next week.

guscarreon commented on exchange/exchangetest/request-multi-bidders-debug-info.json:
```diff
@@ -65,6 +65,7 @@
           {
             "uri": "appnexusTest.com",
             "requestbody": "appnexusTestRequestBody",
+            "requestheaders": null,
```
Can we create a test were `"requestheaders"` is not null?

ShriprasadM replied:
@guscarreon : I have modified _request-multi-bidders-debug-info.json_ as per https://github.com/prebid/prebid-server/pull/1659/commits/92d3b49c3eb18db8e14e30743629b2a4d281b8ba 

_Changes_ 

1. Modified outgoing mock request for appnexus, to send some request header information.

2.Modified sample mock response such that `ext.debug.httpcalls.appnexus.requestheaders` will return the information of passed request headers

guscarreon reviewed (changes requested):
Looks pretty good. Could we create an additional JSON test where `"requestheaders"` is not null? Say a file probably named `exchange/exchangetest/request-debug-info-with-headers.json` where we can assert that the request headers where correctly marshalled into the response.

hhhjort reviewed (dismissed):
Looks good to me.

SyntaxNode commented on exchange/bidder_test.go:
```diff
@@ -886,6 +886,9 @@ func TestBadRequestLogging(t *testing.T) {
 	if ext.Status != 0 {
 		t.Errorf("The Status code should be 0. Got %d", ext.Status)
 	}
+	if nil != ext.RequestHeaders || len(ext.RequestHeaders) > 0 {
+		t.Errorf("The request headers should be empty. Got %s", ext.RequestHeaders)
+	}
```
The `len` operator gracefully handles `nil`. I think you'll be fine with just `len(ext.RequestHeaders) > 0`.

guscarreon replied:
> The `len` operator gracefully handles `nil`. I think you'll be fine with just `len(ext.RequestHeaders) > 0`.

As @SyntaxNode points out, we could do without the first check in the if statement:
```
  -     if nil != ext.RequestHeaders || len(ext.RequestHeaders) > 0 {
  +     if len(ext.RequestHeaders) > 0 {
		t.Errorf("The request headers should be empty. Got %s", ext.RequestHeaders)
	}
```

ShriprasadM replied:
Removed  `nil != ext.RequestHeaders` check

guscarreon reviewed (changes requested):
Looks pretty good @ShriprasadM . I believe the only thing pending is @SyntaxNode 's comment on `exchange/bidder_test.go` line 889

ShriprasadM commented:
@SyntaxNode : I have addressed the review comment

I have also moved `Requestheaders` declaration next to `RequestBody`. Here is the diff - https://github.com/prebid/prebid-server/pull/1659/files#diff-3c1ac989dd21a327ea120cbd84a4d7484528ba9be45903abac53efb283a3f115

@guscarreon  and @SyntaxNode : Please review and merge
@hhhjort : Thanks for review

ShriprasadM commented on openrtb_ext/response.go:
```diff
@@ -55,10 +55,11 @@ type ExtBidderError struct {
-	Status       int    `json:"status"`
+	Uri            string              `json:"uri"`
+	RequestBody    string              `json:"requestbody"`
+	RequestHeaders map[string][]string `json:"requestheaders"`
```
Moved `RequestHeaders` next to `RequestBody`

ShriprasadM commented:
@SyntaxNode : Can you review and approve?

guscarreon reviewed (approved):
LGTM

hhhjort reviewed (approved):
LGTM

ShriprasadM commented:
Thanks @SyntaxNode , @guscarreon  and @hhhjort  

Do we need any documentation update around it ?
